### ISSUE TITLE
ext4: 

### ISSUE THREAD

**Issue by joachimmetz** [sleuthkit/sleuthkit#3140]
Tested with https://github.com/sleuthkit/sleuthkit/commit/7e8eee497a7d5c0d86abff0783dcda2fe891f808

Test data generated with https://github.com/dfirlabs/ext-specimens, manually changed logical block numbers of `/testdir1/TestFile2` to reverse order

Native Linux implementation

```

```
ext4_ext_check_inode:520: inode #15: comm md5sum: pblk 4065 bad header/extent: invalid extent entries - magic f30a, entries 2, max 4(4), depth 0(0)
```

**Comment by simsong** [comment]
Neat. But what is correct behavior? Is it refusing to read the data, or recovering the data? If TSK refused to read, how would the data be read?

**Comment by joachimmetz** [comment]
The native implementation refuses to read the data. For sleuthkit it is probably more suitable to flag/tag the data stream as recovered.

**Comment by simsong** [comment]
I don’t know of any way that we have a tagging the data, other than creating a fake directory entry and returning it in another stream.

This is something that @bcarrier and @jayaramcs need to weigh in on.

**Comment by joachimmetz** [comment]
> I don’t know of any way that we have a tagging the data, other than creating a fake directory entry and returning it in another stream.

accessing inode and/or directory entry (file metadata) works fine, it's the data stream that is considered "corrupt" maybe having a value/function to get a status flag from the corresponding struct in the sleuthkit?

**Comment by simsong** [comment]
> > I don’t know of any way that we have a tagging the data, other than creating a fake directory entry and returning it in another stream.
> 
> accessing inode and/or directory entry (file metadata) works fine, it's the data stream that is considered "corrupt" maybe having a value/function to get a status flag from the corresponding struct in the sleuthkit?

Good idea, but it seems like a lot of work to create the API, then to create the user interface, then to update the documentation, then to inform the user community...

**Comment by joachimmetz** [comment]
you'll need to inform your users anyway give now the implementation misrepresents a error-condition as a non-error-condition and an user could interpret result.

**Comment by msuhanov** [comment]
> you'll need to inform your users anyway given now the implementation misrepresents a error-condition as a non-error-condition and an user could misinterpret the results.

There should be thousands of such cases with all file systems supported. Like out-of-order entries in the NTFS $INDEX_ALLOCATION record, FAT directory entries starting with a null byte (Windows treats them as "end of directory", Linux treats them as "unallocated entries to skip"), or other stuff, especially "manually changed".

Do other Ext4 implementations (like in the GRUB boot manager) also report this corruption?

**Comment by simsong** [comment]
I think that this and other significant issues should be noted, but that we should not even consider addressing them for the next release. I have added a tag called 'correctness' to this. I think that we should tag other issues that are also issues of correctness.

**Comment by joachimmetz** [comment]
> Do other Ext4 implementations (like in the GRUB boot manager) also report this corruption?

Why does NTFS or other ext implementations matter? Isn't this is the sleuthkit issue tracker? Isn't this issue about ext4? The manual change is only a quick way to reproduce the corruption scenario. 

The current behavior of the sleuthkit is to issue a warning for various format corruption cases, e.g. overlapping extents. hence it would be consistent to do the same for this case.

> I think that this and other significant issues should be noted

Thanks much appreciated. If you call them correctness, UX or something else if fine with me. The prioritization what (not) goes into the next release is up to you, I personally would prioritize the memory safety issues first, then the correctness, then new features.

**Comment by msuhanov** [comment]
> Why does NTFS or other ext implementations matter? Isn't this is the sleuthkit issue tracker? Isn't this issue about ext4? The manual change is only a quick way to reproduce the corruption scenario.

Because there are thousands of similar issues (another example is an Ext4 image containing a directory with two files sharing the same name). File system drivers (and, in general, file system parsers) aren't equal in such edge cases.

If it's possible to craft a file system image containing a file that is readable by GRUB (or 7-Zip, or any other tool) but unreadable (causing a read error) by Linux, then many users might want (or even expect) a forensic parser to read that file (cf. a malicious APK file that can't be extracted using usual tools like unzip, but can be installed on Android).

**Comment by joachimmetz** [comment]
Yes I understand the general case, what does that have to do with this issue?

Can you also elaborate what so you specifically mean with "forensic parser"? Wouldn't the fact that a regular user could not read the contents of a file be a relevant fact in a legal context?

**Comment by msuhanov** [comment]
> Can you also elaborate what so you specifically mean with "forensic parser"?

A file system parser (either a library or a tool) that is intended to be used in digital forensics, incident response, and, to some (limited) extent, data recovery.

> Wouldn't the fact that a regular user could not read the contents of a file be a relevant fact in a legal context?

A malformed Ext4 image can be delivered to a "regular" Windows user as a file and opened by them via 7-Zip (which implements many relaxed file system checks). Also, native file system implementations can treat the same corruption differently depending on their versions (i.e., the same format violation can be ignored by an older file system driver).

So, this fact is relevant, but it can't be proven until the whole situation is evaluated. A known format violation alone doesn't prove that a user couldn't read the contents of a given file.

---

Some file system corruption scenarios are reachable on properly functioning hardware due to software bugs (one example is here: https://github.com/exfatprogs/exfatprogs/issues/261#issuecomment-2203014205). So, it might be better to introduce the "recovered" flag to mark files, directories, and whole file systems (like in that checksum case with exFAT), as suggested before:

> For sleuthkit it is probably more suitable to flag/tag the data stream as recovered.

**Comment by joachimmetz** [comment]
> A file system parser (either a library or a tool) that is intended to be used in digital forensics, incident response, and, to some (limited) extent, data recovery.

* intended by whom? by the author?
* would that make 7-Zip and GRUB boot manager non-"forensic parsers" ? Neither of them seem to claim that are intended for digital forensics purposes? What's the relevance of specially calling them out then?

> So, this fact is relevant, but it can't be proven until the whole situation is evaluated. A known format violation alone doesn't prove that a user couldn't read the contents of a given file.

Agree hence I framed it as fact, and not as evidence or proof

>  So, it might be better to introduce the "recovered" flag to mark files, directories

So you are agreeing with me? Given that is what I said in https://github.com/sleuthkit/sleuthkit/issues/3140#issuecomment-2558374971 and in lack of this feature, the closest thing is the warning message the Sleuthkit tools issue (which is a notification to the user see https://github.com/sleuthkit/sleuthkit/issues/3140#issuecomment-2558586107)

**Comment by msuhanov** [comment]
> What's the relevance of specially calling them out then?

Both of them are widely used to parse file systems. Of course, GRUB and 7-Zip aren't reference implementations, but they are so broadly used, so, in my opinion, forensic parsers should be aligned here.

> So you are agreeing with me?

Maybe. The "recovered" flag would be useful, but file system parsers aren't equal, so the absence of this flag would never mean that the file system is "good" (not malformed or corrupted according to the "official" implementation). And there are thousands of possible corruption scenarios...

**Comment by joachimmetz** [comment]
> in my opinion, forensic parsers should be aligned here.

That appears one of the crux in this issue (and the broader DFIR field), it is mostly opinion based instead of principle based.

> Maybe. The "recovered" flag would be useful, but file system parsers aren't equal, so the absence of this flag would never mean that the file system is "good" (not malformed or corrupted according to the "official" implementation). And there are thousands of possible corruption scenarios...

Note that is issue is not about having the "ultimate digital forensics parsing solution" but regarding consistency in the Sleuthkit tooling. Also see https://github.com/sleuthkit/sleuthkit/issues/2855 for where the tooling incorrectly notify the user about a file system being corrupt.
